### PR TITLE
expression: introduce `StaticEvalContext` to provide context when evaluating expressions

### LINKED ISSUE #52630: introduce `StaticEvalContext` to provide context when evaluating expressions

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #52630

### What changed and how does it work?

- Move the definition of `SQLWarn` to `pkg/util/context`. Add added some methods to `WarnHandler`. The old interface is renamed to `WarnAppender`.
- Introduce `StaticWarnHandler` to provide a thread-safe implementation.
- Introduce `StaticEvalContext` to  provide a static implementation of `EvalContext`. We also provides a function `NewStaticEvalContext` to create a new instance. For example:

```
// create a context with specified timezone and type flags.
ctx = NewStaticEvalContext(
    WithLocation(time.Local),
    WithTypeFlags(types.FlagAllowNegativeToUnsigned|types.FlagSkipASCIICheck),
    // ....
)
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by YangKeao — APPROVED** [review]
LGTM

**Review by lance6716 — COMMENTED** [review]
/approve

**Review by hawkingrei — APPROVED** [review]

File: pkg/expression/contextstatic/evalctx.go [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+}()
+
+// StaticEvalContext implements `EvalContext` to provide a static context for expression evaluation.
+type StaticEvalContext struct {
+	id uint64
+	staticEvalCtxState
+}
+
+// NewStaticEvalContext creates a new `StaticEvalContext` with the given options.
+func NewStaticEvalContext(opt ...StaticEvalCtxOption) *StaticEvalContext {
```
Comment (zimulala) [inline pkg/expression/contextstatic/evalctx.go:209]: This and "NewStaticWarnHandlerWithHandler" seems to be no use. Is there any other PR that will use them later?

Comment (lcwangchao) [reply pkg/expression/contextstatic/evalctx.go:209]: These methods is going to be used in two scenes:

1. replace `mock.Context` to do some tests or evaluate expressions in production code.
2. "Clone" a `SessionEvalContext` for cursor fetch.

**Review by zimulala — APPROVED** [review]

File: pkg/expression/contextstatic/evalctx.go [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+	}
+
+	tm = tm.In(loc)
+	t.time.Store(&tm)
+	return
+}
+
+// staticEvalCtxState is the internal state for `StaticEvalContext`.
+// We make it as a standalone private struct here to make sure `StaticEvalCtxOption` can only be called in constructor.
+type staticEvalCtxState struct {
```
Comment (windtalker) [inline pkg/expression/contextstatic/evalctx.go:70]: Can you give some explaination about why the `staticEvalCtxState` is named by `static`?

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**